### DISCUSSION TITLE
Code on RPi Pico

### DISCUSSION THREAD

**Discussion by ebolisa** [n-elia/MAX30102-MicroPython-driver#5] [Show and tell]
Hi, 
Testing the library code (no changes made) on a Pico board and getting pretty much a linear output.
What is possible wrong?
TIA

![Cap](https://user-images.githubusercontent.com/17126140/118961919-20ea6600-b965-11eb-85ad-1ff1c9f8c218.PNG)

**Comment by n-elia** [comment]
Hi ebolisa!
First of all, let me know which is your particular board, and which is your sensor (the original Maxim, or a clone?).
Then, please try to scale the y-axis, e.g., plot the _Columna B_ only with an y-axis ranging from 18000 to 22000.
Thank you!

**Comment by ebolisa** [comment]
Thanks for getting back to me.
I'm using the Raspberry Pi Pico and the sensor board may be a clone: https://www.pantechsolutions.net/max30102-pulse-oximeter-heart-rate-sensor-module-i2c-interface.
As for the graph, I'm using OpenOffice and I don't control it that well so I'm attaching some raw data.
[data.csv](https://github.com/n-elia/MAX30102-MicroPython-driver/files/6517216/data.csv)
![Cap](https://user-images.githubusercontent.com/17126140/119010067-fb288580-b993-11eb-86dd-d422bbc0addc.PNG)
EDIT:
Updated sensor's link.

**Comment by ebolisa** [comment]
Ok, Made progress. Had a crash course on Calc.
But how do I output in a range from 0 to 100% for Oxy. And < 200 for HR?
TIA
![Cap](https://user-images.githubusercontent.com/17126140/119018407-55c5df80-b99c-11eb-979b-6981bc1b5dac.PNG)

**Comment by n-elia** [comment]
You're very welcome.

So, as I could see in your link, this is a MAX30100, so I am not sure that this library will work. However, we can further investigate. Let's check something:

1. The csv that you posted contains 26 points only. if I remember well, with the default settings, the sensor will provide 50 samples per second, therefore you should plot at least 200 samples (4 seconds) to see an heart signal that makes sense. 
2. You can easily check the acquisition frequency with the code proposed in the readme; please make sure to have at least 50Hz
```
# Get the estimated acquisition rate
acquisition_rate = sensor.getAcquisitionFrequency()
```
3. How are you logging data? Is it fast enough to get all the samples?

Let me know!

**Comment by n-elia** [comment]
Nice to see that it works well!

You have to do some math to obtain SpO2 and HR. I put some hints in the last lines of the readme, [here](https://github.com/n-elia/MAX30102-MicroPython-driver#other-things-that-its-worth-mentioning).

Please share your code if you manage to perform those computations locally, I am interested, too!

**Comment by ebolisa** [comment]
Sorry, I posted the wrong sensor's link so, I updated my post.

**Comment by ebolisa** [comment]
Will do. Thanks again.

**Reply by Abhishekcr448** [reply]
Could you please share your code on this? Because I have been trying to configure it to raspberry pi pico for the past two weeks, I am still not able to find the correct way.
I am trying it on Thonny.
I would be able to submit my project.
Also please mention steps, if you have any.

**Comment by n-elia** [comment]
Fine! Then, I'm going to close this issue.

Please feel free to open a topic in the "Discussion" section if you want to look for help for implementing HR/SPO2 evaluation, or if you want to share some progresses.

Bye,
Nicola

**Reply by Abhishekcr448** [reply]
Could you please share a code on this? Because I have been trying to configure it to raspberry pi pico for the past two weeks, I am still not able to find the correct way.
I am trying it on Thonny.
I would be able to submit my project.
Also please mention steps, if you have any

**Reply by n-elia** [reply]
> Could you please share a code on this? Because I have been trying to configure it to raspberry pi pico for the past two weeks, I am still not able to find the correct way. I am trying it on Thonny. I would be able to submit my project. Also please mention steps, if you have any

Hi, have you changed the pins to your board's I2C pins (16,17) ? Feel free to open a new discussion, attaching your code and the steps you made so far!

**Reply by Abhishekcr448** [reply]
So I downloaded the module from [Here](https://github.com/n-elia/MAX30102-MicroPython-driver) and just copied the files 
1)max30102.py and 2) circular_buffer.py into Thonny and then wrote this code in the main.py
![max30102 img](https://user-images.githubusercontent.com/106028156/169857820-1cfd0ac5-d3eb-4e1e-a8c1-c1e896c28ca2.PNG)
as you can see in the shell there are no errors, but there are also no values that I expected.
I guess because I am trying it on Thonny, it is not able to plot the graph, could you please tell me the way that I could directly print the corresponding values in the shell
I just need the list of values every 0.5 sec.

**Reply by n-elia** [reply]
Hi, 
what do you mean with "copied into Thonny"?
Please try to upload files into the board such that  the board content is as follows:
```
/
└── root.py
└── main.py
└── lib
    └── circular_buffer.py
    └── max30102.py
```
Then, you will just need to connect to your board terminal and press the reset button to run the main.py script.

Also please check that you're using the correct pins (16, 17 according to your code), and use this code as `main.py`:
```python
# main.py
from machine import sleep, SoftI2C
from utime import ticks_diff, ticks_ms

from max30102 import MAX30102

if __name__ == '__main__':
    i2c = SoftI2C(scl=machine.Pin(17),
                          sda=machine.Pin(16),
                          freq=400000)
    sensor = MAX30102(i2cHexAddress=0x57, i2c=i2c)

    print("Setting up sensor with default configuration.", '\n')
    sensor.setup_sensor()
    # Set the sample rate to 800: 800 samples/s are collected by the sensor
    sensor.set_sample_rate(800)
    # Set the number of samples to be averaged per each reading
    sensor.set_fifo_average(8)
    sleep(1)

    print("Starting data acquisition from RED & IR registers...", '\n')
    sleep(1)

    while True:
        sensor.check()

        if sensor.available():
            red_reading = sensor.pop_red_from_storage()
            ir_reading = sensor.pop_ir_from_storage()

            # Print the acquired data (so that it can be plotted with a Serial Plotter)
            print(red_reading, ",", ir_reading)
```

**Reply by Abhishekcr448** [reply]
Did exactly as what you said,
![max30102 img](https://user-images.githubusercontent.com/106028156/169883487-321db0e5-117b-4c77-9a7c-aee34de14c33.PNG)

but still nothing changed. As you can see there is no output.
There is also no light in LED when I start the code, does that mean that my chip could be defected. But the sensor could be detected, I can detect it's address.
At this point I am not getting where is the issue.

**Reply by n-elia** [reply]
Uhm.... me too. Have you checked the power supply wiring of your MAX30102 sensor?

**Reply by Abhishekcr448** [reply]
Where do I have to connect the INT, RD, and IRD pins? When I tried connecting the RD pin to the ground or 3V3(out) the light turned on (for the first time).

**Reply by n-elia** [reply]
Hi, this library works with MAX30102/MAX30105 sensors on I2C breakout boards.
Therefore you should have 2 pins for power supply and 2 pins for I2C communication.
At the end of the day, you should connect 4 wires:
- GND: to board GND pin
- Vcc (or Vin): to board 3.3V pin
- SDA: to board pin 16
- SCL: to board pin 17

I just released a new version v0.3.5 with an update example script which offers better error handling. 
Please use the new example script, and let me know. You should see at least an error message.

Edit: after saving files with Thonny IDE, remember to reset the board using the onboard physical reset button (Ctrl-D is not enough).

**Reply by n-elia** [reply]
> Where do I have to connect the INT, RD, and IRD pins? When I tried connecting the RD pin to the ground or 3V3(out) the light turned on (for the first time).

INT is used to provide hardware interrupts to the main board when measurements are available. Other two pins are drivers for red and IR leds. You do not need those 3 pins to be connected when using this library.

**Reply by Abhishekcr448** [reply]
Ok, So I used your new code followed your steps by restarting the board changed the values of sda and scl from 22, 21 to 16,17 respectively and then got this error:
![max30102 img](https://user-images.githubusercontent.com/106028156/170084783-8ea865a0-2556-46a8-b8ae-d627361dbfbc.PNG)
When I restarted the board (with the Bootsel button) and run the main.py file it gave me this error: _Error while reading from I2C bus: OSError code 116_
So to check that my board was able to detect the sensor, I used this simple code to find it:
`from machine import Pin, I2C
sda=Pin(16)
scl=Pin(17)
i2c=I2C(0,sda=sda, scl=scl, freq=400000)
print('Scan i2c bus...')
devices = i2c.scan()
if len(devices) == 0:
    print("No i2c device !")
else:
    print('i2c devices found:',len(devices))
for device in devices:
    print("Decimal address: ",device," | Hexa address: ",hex(device))`
As you can see in the image it successfully detected the sensor and gave the address in the shell
So after that, when I again started the main.py file it gave me this error:_Sensor not found on I2C bus._ this time there was no _OSError code 116_

Now I don't know why it says sensor not found, but what is meant by OSError code 116? and is there something missing or done wrong by me?

**Reply by n-elia** [reply]
Uhm... when you upload your main.py and boot.py files, they should be executed automatically as soon as you release the 'reset' button.
Please do not run scripts from Thonny, instead (see [here](https://docs.micropython.org/en/latest/wipy/tutorial/reset.html?highlight=main%20py#reset-and-boot-modes) for more details) let the board hard reset by pressing the reset button (bootsel, in your case). Use Thonny to upload files and to open the remote terminal only.

Also, please remove other scripts onboard (delete everything except the files needed to run the example).

Please retry as suggested and let me know.

PS: the sensor address that you found is correct (0x57), so wiring and pin numbers are OK. The OSError you're receiving is new to me (it is not documentend by [docs](https://mpython.readthedocs.io/en/latest/library/pythonStd/uerrno.html)), but is should be related to files (thus, it should be solved if you follow the previously mentioned suggestions).

Edit: I saw somewhere that people is having errors with I2C frequency. A further attempt could be to put 200000 instead of 400000.

**Reply by Abhishekcr448** [reply]
Thanks for the information, but the error is the same: _Error occurred while sensor initialization: Error while reading from I2C bus: OSError code 116_. 
I have read that the OSError is the timeout error, which means it could be either due to loose connections or the code is not persevering any data.
Also, the freq=400000 is fine, and if you don't mention it, then it would take the highest default value. (which is 400000 for pico)
You should ask others too, if it's working for them there could be a fault in my sensor.

**Reply by n-elia** [reply]
Sorry to hear that. I think that it's your sensor's fault. The library works well and we (me and the community) tested it on many devices.

I bet that you are not using a genuine Maxim sensor. It happens frequently thath chinese clones are faulty.

The `SoftI2C` instance uses `readfrom()` method for reading from I2C bus, targeting the sensor address (0x57). The OSError that you see is thrown by this method. This means that the board is unable to "talk" with your sensor on the I2C bus.

**Reply by Abhishekcr448** [reply]
I have decided to buy a new sensor. Do you have any recommendations on some specific version or any e-commerce site where I could get an original one?

**Reply by n-elia** [reply]
Hi, I would buy the genuine Maxim MAX30102 breakout ([MAXREFDES117#](https://www.mouser.it/ProductDetail/Maxim-Integrated/MAXREFDES117?qs=L5CenAjuTY7RgAB22JU8Sw%3D%3D)) or, alternatively, a non-Maxim breakout such as the Pimoroni ([PIM438](https://www.digikey.ee/catalog/en/partgroup/max30105-breakout-heart-rate-oximeter-smoke-sensor-board/88533)) and the DFRobot Fermion ([SEN0344](https://www.dfrobot.com/product-2164.html)).
The Pimoroni that I mentioned is based on MAX30105. Thus, it offers the green LED, too (better for SpO2 estimation).
The e-commerce choice depends on where you live in. Here in Italy, I usually rely on Mouser, Dfrobot, Digikey, Amazon.

edit: please keep us updated :smiley:

**Reply by Abhishekcr448** [reply]
Could I buy this? 
![m](https://user-images.githubusercontent.com/106028156/170810654-219c52b7-e878-40cb-ae5b-d3d6b564de9f.jpg)
I guess this should work.

**Reply by n-elia** [reply]
For sure, but I cannot guarantee that it has the very same performance as branded ones.

**Reply by Abhishekcr448** [reply]
I have bought the new chip, and this time it's showing a different error.
`Error occurred while sensor initialization: Sensor not found on I2C bus.`
But I have done the test 'sensor_test' and it's detecting the sensor. But the 'example' code is not detecting it. What could be the issue?
I have connected it to 16 and 17 pins, and nothing is changed. 
I guess there could be some error in the code because I bought a new pressure sensor from the same site and it's working fine.
I guess there is something in the code which is not reading the details of the chip correctly.
There could be something wrong here:
`try:
            self._i2c.readfrom(self._address, 1)
        except OSError as error:
            if error.errno == uerrno.ENODEV:
                raise RuntimeError("Sensor not found on I2C bus.")`

**Reply by n-elia** [reply]
Hi, please share your full code (`main.py`) and the connection scheme.

**Reply by Abhishekcr448** [reply]
Ok so, the main.py is nothing but your example code, in which I have changed the pins to 16 and 17 for SDA and SCL respectively. 
Firstly I checked whether pico is able to detect the sensor, and for that did this simple sensor test:
![sensor_test](https://user-images.githubusercontent.com/106028156/172366106-82094ad5-baf9-445a-9f07-8d3eeb7b21c4.PNG)
As you can see it is able to detect the sensor, on 16 and 17 positionings.

Then I did run your main.py file, and it showed this error:
![max_test](https://user-images.githubusercontent.com/106028156/172366366-fbc33521-775e-493b-b0f5-70a473f8651d.PNG)
As you can see, it says 'Sensor not found on I2C bus'

Both the codes are clearly showing the changes I made (in pin no.)
I took little time to reply because I checked whether it could run on a different code. So I checked it with Arduino Uno with a Sparkfun code and it was successful. So my new sensor is working fine. I cannot use that code, because I don't know how to compile a C code in Micropython. Thus I need a code in python for it.

**Reply by n-elia** [reply]
I can't find errors.
The library works as expected, I am testing it on a Esp32 board before releasing.

I noticed that you're still running main.py from thonny. Instead, you should just reset the board and let it run the boot and main.py automatically.
Edit: when you edit and save the main.py, be sure to save it on the board and not on your PC local directory.

Also please try lowering the I2C frequency from 400000 to 100000 and let me know.

**Reply by Abhishekcr448** [reply]
Sorry, there is no difference after changing the frequency. And how could I be able to see the print statements without using Thonny?
Also, I don't understand how could Sparkfun's code can run the chip and not this.
If you look there is an error in sensor detection. so it's not able to get the correct address.  There is an error in the try statement: `try:
            self._i2c.readfrom(self._address, 1)
        except OSError as error:
            if error.errno == uerrno.ENODEV:
                raise RuntimeError("Sensor not found on I2C bus.")`
It is not able to readfrom i2c and I don't know why

**Reply by n-elia** [reply]
Hi, I'm sorry but you're missing the basic knowledge. The error that you are seeing has been put there by me.

From [Get Started with MicroPython on Raspberry Pi Pico](https://store.rpipress.cc/products/get-started-with-micropython-on-raspberry-pi-pico):
> When MicroPython finds a file called main.py in its file system, it runs that automatically every time it’s powered on or reset – without you having to click Run.  In Thonny, after stopping the program if running, click the File menu then Save As. Click ‘Raspberry Pi Pico’ in the pop-up that appears, then type ‘main.py’ as the file name before clicking Save.

**You just need Thonny to upload .py files onto the board, and then use Thonny for reading the serial output. You should NOT execute the file manually.** The underlying motivation is that I2C connection needs a hard reset.

Let me give you an example.
I have a ESP32 board and a MAX30102 which should be the very same of yours.
![image](https://user-images.githubusercontent.com/62206011/172397105-ef27ac11-0cd4-481a-bb5d-8d71a7011be9.png)
I open Thonny, and I upload files on the board. Then, If I issue Ctrl+D on the board shell, Micropython does a soft reboot and therefore the `boot.py` and `main.py` are executed. As you can see, I get the `Sensor not found on I2C bus.` error. If I reboot (i.e. power off and power on) my board, the script starts running fine. Thus, I'm using Thonny just for reading the serial output.
![image](https://user-images.githubusercontent.com/62206011/172396635-efb2b593-d5f8-42f5-af22-9ce26ec7f322.png)

**Reply by Abhishekcr448** [reply]
FINALLY, found the error. The error was in your code.
I added one single line after declaring i2c which is `devices = i2c.scan()`
Here is the image of where I exactly added it:
![final](https://user-images.githubusercontent.com/106028156/172425150-5837ddc0-44cc-4087-b3d4-dc9548eddc68.PNG)
I didn't even share that `devices` variable into the main max30102 module, and it's still working. I guess the scan function was remaining. My solution might not be very sofesticated. I recommend you have a look at this change in code. If it's working for me it could mean something.

I know how the main.py works, but if that file has been initialised manually still it should give the desired values. (That's how a code works)

And lastly Thank you so much for helping me all these days without any hesitation. You were like a guide to me.
People like you are a motivation for me to code something on an open-source platform and help others.
I will surely share things about my final product, either img or video along with the code with you.
Thank You🙏

**Reply by n-elia** [reply]
Never surrender! Many thanks for your kind words.

It does not make sense to me to perform a i2c scan without using the returned `devices` list! I'll investigate on that and eventually get back on this page.

i look forward to see your final results.

edit: please consider that soft reset clears the Micropython interpreter, but the underlying OS, the network stack and the hardware modules are not reset. For instance, if you open sockets and do a soft reset, you'll get an error to use these sockets again. So, for sure the same code should bring to the same results, but a soft reset cannot guarantee you to have deterministic results.

**Reply by Abhishekcr448** [reply]
I was looking for finding heart rate and spo2 by your given references, but I am blank. I am not able to use their code in this code. 
I am not able to find a specific formula for this.
Could you please give the formula to find these by the red and IR readings, or
Could you please write a code which could find heart rate and spo2 on your module?

**Reply by n-elia** [reply]
> I was looking for finding heart rate and spo2 by your given references, but I am blank. I am not able to use their code in this code. I am not able to find a specific formula for this. Could you please give the formula to find these by the red and IR readings, or Could you please write a code which could find heart rate and spo2 on your module?

Hi, I'm sorry but I have no free time to work on this. Eventually, I would be happy to include your findings into  a new example script.

**Reply by n-elia** [reply]
Hi, @Abhishekcr448 
I released a new version that should work with your board, without manually adding `i2c.scan()`.
Bye,

**Reply by Darkai87** [reply]
Hello n-elia, I'm currently doing a vital signs monitor group project as part of my role in the group and was wondering how I would go about coding the MAX30102 with the raspberry pi pico to calculate heartrate and SPO2? I've been struggling for the past two to three weeks and still haven't seen any progress seen you can't import hrcalc nor numpy to code the calculations necessary to measure and display heartrate and SPO2 readings. I know the test.py you can obtain the ADC values, but I would really appreciate the help! Thanks.

**Reply by n-elia** [reply]
> Hello n-elia, I'm currently doing a vital signs monitor group project as part of my role in the group and was wondering how I would go about coding the MAX30102 with the raspberry pi pico to calculate heartrate and SPO2? I've been struggling for the past two to three weeks and still haven't seen any progress seen you can't import hrcalc nor numpy to code the calculations necessary to measure and display heartrate and SPO2 readings. I know the test.py you can obtain the ADC values, but I would really appreciate the help! Thanks.

Hi, to implement a pulseoximeter with a RPi you will need:
- Raspberry Pi board
- Max30102 or MAX30105 sensor, properly connected to the i2c bus of raspberry pi
- Python library (not my library, that is for micropython), for instance [this](https://github.com/pimoroni/max30105-python).
- A script that takes raw data from the sensor, with the proper settings, and does the math for converting to HR and SPO2 values. For instance, you can use [this](https://python-heart-rate-analysis-toolkit.readthedocs.io/en/latest/).

**Comment by Darkai87** [comment]
Okay yes I've acquired all those now I'm getting an error in my program for line 5 in the main.py "m = max30102.MAX30102()" and line 55 in the __init__ module saying "self.bus = smbus.SMBus(self.channel)". I'm now wondering what the problem could be my microcontroller is the raspberry pi zero which still has all necessary libraries in order for the MAX30102 to run.

**Reply by n-elia** [reply]
Hi,
Let's do a basic check: did you enable I2C using raspi-config?
If not, try to open a terminal and run `sudo raspi-config`. Then, select "Interfacing options"->"I2C" choose and hit Enter, then go to Finish and reboot your raspberry.
